### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by kripton**
So, here it is: The thing that I would like to call v1.0.0.
As the title suggests, it's a complete re-write of what was there before, splitting it up in proper C++ classes. There are still a LOT of TODOs in the code, but it is at a state that I would definitely call "usable".

Main features:
* Status LEDs (8 WS2812 LEDs), dimmable in case they are too bright
* Web interface (dongle enumerates as an NCM network adapter and hosts its own web server for configuration and status)
* Logging via USB serial interface or accessible via web server (logs are buffered to some extent)
* Detects connected IO boards and their type (connector and port direction). However, currently treats all boards as 4 outputs
* Configurations are stored on one of the IO Boards or the main board and managed on the WebUI
* Optional wireless module is detected and used for wireless DMX (currently TX and RX is implemented, Mesh is still missing). I've successfully transferred four DMX universes on one radio channel without noticeable latency
* Main method to get DMX data from the host PC to the board is either ArtNet or sACN (the board speaks both). A "native USB protocol" is designed and integrated into the dongle (the same protocol as used for the wireless DMX, including data compression) but no PC software has yet been adapted to use it. Main advantages of "network"-based protocols:
  * Super-easy hotplugging: Dongle starts processing data from host in less than one second after plugging in. libusb-based dongles and hotplugging (also in the sense of disconnecting and reconnecting the dongle while it's running) is quite some effort
  * Easy to debug using tools such as Wireshark
  * Already supported by a great number of lighting control software
*  The dongle has 24 internal buffers that can be patched pretty flexibly. In this version, it's simply hardcoded:
   * buffer -> local output for the fist 16 buffers +
   * buffer -> wireless for the first 4 buffers +
   * wireless -> buffer (4 universes incoming go to buffers 4 to 7 = outputs 0-3 on second IO board). So it can be used as a standalone Wireless -> DMX receiver without host PC.

Screenshots of web interface can be provided. However, everything required to test it is a Raspberry Pi Pico board, one microUSB-Cable and the firmware file :). Load the firmware onto the board, reboot it. It will/should show up as a network card (works on Linux, macOS and Windows hosts) and assign an IPv4-address to your machine. Then, open a web browser and enter the dongle's IP there. The IP is `169.254.X.1` (link-local addressing) and can be found out using `dmesg` on Linux or hovering over the "USB" icon in the task bar on Windows. Mac users should be able to see the IP in the network settings but since I don't own a Mac I didn't test that.

I plan to run a first "real" show with that interface (maybe even with wireless ;) in December. If the Corona-situation allows it ...

### CODE REVIEW

**Comment by kripton** [comment]
Oh, fixes #15 ;)

File: wwwsrc/src/Home.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+                <div className="row">
+                    <div className="col">
+                        &nbsp;
+                    </div>
+                </div>
+
+                <div className="row">
+                    <div className="col">
+                        Running configuration and connected Io boards:
```
Comment (peternewman) [inline wwwsrc/src/Home.js:153]:
```suggestion
                        Running configuration and connected IO boards:
```

**Comment by peternewman** [comment]
> By the way, I'm not expecting a line-by-line code review. This is more to see if the direction this is going is fine for the other OLP-devs :)

Can I say this is amazing that you've done the whole stack single handed!

I've only played with the web UI so far but it looks great!

> Especially if I can go ahead requesting USB PID 0x2040 as stated in https://github.com/OpenLightingProject/rp2040-dongle/pull/16/files#diff-40b405bc8e91746f1ee25c24a2cf6e409689d83f380d95094b60cfbf9b359360R34

I'd imagine so, but please can you put a comment on the line you're referring too, the link doesn't work for me.

**Review by peternewman — CHANGES_REQUESTED** [review]
Typo nit!

**Comment by kripton** [comment]
> Can I say this is amazing that you've done the whole stack single handed!

Thank you! I did some web-page-projects in my day-to-day-job but the main challenge here is to make sure that the firmware (compiled C++ code) plus the complete web application including assets (icons, pictures, ...) fit into the 2MB flash of the Pico board. This is why I chose React over the other available web frameworks :)

> I've only played with the web UI so far but it looks great!

Great that it was easy for you to get it up and running to that point! To which host OS did you connect your Pico board? Linux. macOS or Windows?
Without any additional hardware, you could do the following things:
* Send ArtNet or sACN to the board's IP address (sACN should also work with multicast) and see if the values change on the WebUI's "Console" page
* Use the "Config" page to edit the board's name or IP-address, pressing the "Save to base board" button to save it to the Pico's Flash storage and then rebooting (= unplugging + re-plugging) it. That should make the Status LEDs on the "Home" page show a different status (since it now has a "real", valid config instead of the fallback one). And of course the modified board name and IP should be reflected on the "Home" and "Config" pages as well

> > Especially if I can go ahead requesting USB PID 0x2040 as stated in https://github.com/OpenLightingProject/rp2040-dongle/pull/16/files#diff-40b405bc8e91746f1ee25c24a2cf6e409689d83f380d95094b60cfbf9b359360R34
> 
> I'd imagine so, but please can you put a comment on the line you're referring too, the link doesn't work for me.

Oh, yeah, that link is broken. I was referring to the file `src/usb_descriptors.c`, line 34 (https://github.com/kripton/rp2040-dongle/blob/6778ae6d7bcc06281f8af6390c11cfbea7ec0c19/src/usb_descriptors.c#L34). The board currently uses the yet unclaimed PID of 0x2040 just for the fun of it (= reference to the RP2040 µC). I could of course also use the one of the JaRule!-Board + 1 as we discussed previously. No preference here.

**Comment by peternewman** [comment]
> > Can I say this is amazing that you've done the whole stack single handed!
> 
> Thank you! I did some web-page-projects in my day-to-day-job but the main challenge here is to make sure that the firmware (compiled C++ code) plus the complete web application including assets (icons, pictures, ...) fit into the 2MB flash of the Pico board. This is why I chose React over the other available web frameworks :)

Yeah that certainly sounds like a challenge! You might want an action that catches such issues as part of the build; the Arduino people do the same thing.

> > I've only played with the web UI so far but it looks great!
> 
> Great that it was easy for you to get it up and running to that point! To which host OS did you connect your Pico board? Linux. macOS or Windows?

A slightly old version of Ubuntu Linux.

> Without any additional hardware, you could do the following things:
> 
>     * Send ArtNet or sACN to the board's IP address (sACN should also work with multicast) and see if the values change on the WebUI's "Console" page. You could also hook up a scope, logic analyser or RS-485 transmitter to the GPIOs of the board (Port 0 is on GPIO 6) to see the DMX data being generated.

I don't think I tried this.

>     * Use the "Config" page to edit the board's name or IP-address, pressing the "Save to base board" button to save it to the Pico's Flash storage and then rebooting (= unplugging + re-plugging) it. That should make the Status LEDs on the "Home" page show a different status (since it now has a "real", valid config instead of the fallback one). And of course the modified board name and IP should be reflected on the "Home" and "Config" pages as well

I think I'd done this already!

> > > Especially if I can go ahead requesting USB PID 0x2040 as stated in https://github.com/OpenLightingProject/rp2040-dongle/pull/16/files#diff-40b405bc8e91746f1ee25c24a2cf6e409689d83f380d95094b60cfbf9b359360R34
> > 
> > I'd imagine so, but please can you put a comment on the line you're referring too, the link doesn't work for me.
> 
> Oh, yeah, that link is broken. I was referring to the file `src/usb_descriptors.c`, line 34 (https://github.com/kripton/rp2040-dongle/blob/6778ae6d7bcc06281f8af6390c11cfbea7ec0c19/src/usb_descriptors.c#L34). The board currently uses the yet unclaimed PID of 0x2040 just for the fun of it (= reference to the RP2040 µC). I could of course also use the one of the JaRule!-Board + 1 as we discussed previously. No preference here.

My gut feeling is this is probably a little niche to be grabbing such a prime PID, especially as a few others haven't taken it either (although they mostly seem to have name based vanity ones instead. The ACE* range is still pretty cool! But ultimately it's your baby so I'll leave it up to you!

**Comment by kripton** [comment]
> > Thank you! I did some web-page-projects in my day-to-day-job but the main challenge here is to make sure that the firmware (compiled C++ code) plus the complete web application including assets (icons, pictures, ...) fit into the 2MB flash of the Pico board. This is why I chose React over the other available web frameworks :)
> 
> Yeah that certainly sounds like a challenge! You might want an action that catches such issues as part of the build; the Arduino people do the same thing.

Could you give me hint here? An action that tells us that the resulting .uf2-file got too large? I looked for the solution in the Arduino-project you mentioned here: https://github.com/earlephilhower/arduino-pico/tree/master/.github/workflows but I couldn't find anything related to binary size.

> > > I've only played with the web UI so far but it looks great!
> > 
> > 
> > Great that it was easy for you to get it up and running to that point! To which host OS did you connect your Pico board? Linux. macOS or Windows?
> 
> A slightly old version of Ubuntu Linux.

Great to hear that it worked out-of-the-box!

> > > > Especially if I can go ahead requesting USB PID 0x2040 as stated in https://github.com/OpenLightingProject/rp2040-dongle/pull/16/files#diff-40b405bc8e91746f1ee25c24a2cf6e409689d83f380d95094b60cfbf9b359360R34
> > > 
> > > 
> > > I'd imagine so, but please can you put a comment on the line you're referring too, the link doesn't work for me.
> > 
> > 
> > Oh, yeah, that link is broken. I was referring to the file `src/usb_descriptors.c`, line 34 (https://github.com/kripton/rp2040-dongle/blob/6778ae6d7bcc06281f8af6390c11cfbea7ec0c19/src/usb_descriptors.c#L34). The board currently uses the yet unclaimed PID of 0x2040 just for the fun of it (= reference to the RP2040 µC). I could of course also use the one of the JaRule!-Board + 1 as we discussed previously. No preference here.
> 
> My gut feeling is this is probably a little niche to be grabbing such a prime PID, especially as a few others haven't taken it either (although they mostly seem to have name based vanity ones instead. The ACE* range is still pretty cool! But ultimately it's your baby so I'll leave it up to you!

Yeah, I just decided to go with the ACE*-range. And just for fun I re-checked on the 0x2040: taken since 2 hours ago ;) I'll send the PR the coming days :+1:

**Pull request merged**